// @material-ui/core/esm/Breadcrumbs/Breadcrumbs.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React from 'react';
import warning from 'warning';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { componentPropType } from '@material-ui/utils';
import withStyles from '../styles/withStyles';
import Typography from '../Typography';
import BreadcrumbCollapsed from './BreadcrumbCollapsed';
import BreadcrumbSeparator from './BreadcrumbSeparator';
var styles = {
  /* Styles applied to the root element. */
  root: {},

  /* Styles applied to the ol element. */
  ol: {
    display: 'flex',
    flexWrap: 'wrap',
    alignItems: 'center',
    padding: 0,
    // Reset
    margin: 0,
    // Reset
    '& li': {
      listStyle: 'none'
    }
  },

  /* Styles applied to the li element. */
  li: {},

  /* Styles applied to the separator element. */
  separator: {}
};

var Breadcrumbs =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Breadcrumbs, _React$Component);

  function Breadcrumbs() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Breadcrumbs);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Breadcrumbs)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      expanded: false
    };

    _this.handleClickExpand = function () {
      _this.setState({
        expanded: true
      });
    };

    return _this;
  }

  _createClass(Breadcrumbs, [{
    key: "insertSeparators",
    value: function insertSeparators(items) {
      var _this2 = this;

      return items.reduce(function (acc, current, index) {
        if (index < items.length - 1) {
          acc = acc.concat(current, React.createElement(BreadcrumbSeparator // eslint-disable-next-line react/no-array-index-key
          , {
            key: "separator-".concat(index),
            className: _this2.props.classes.separator
          }, _this2.props.separator));
        } else {
          acc.push(current);
        }

        return acc;
      }, []);
    }
  }, {
    key: "renderItemsBeforeAndAfter",
    value: function renderItemsBeforeAndAfter(allItems) {
      var _this$props = this.props,
          maxItems = _this$props.maxItems,
          itemsBeforeCollapse = _this$props.itemsBeforeCollapse,
          itemsAfterCollapse = _this$props.itemsAfterCollapse; // This defends against someone passing weird input, to ensure that if all
      // items would be shown anyway, we just show all items without the EllipsisItem

      if (itemsBeforeCollapse + itemsAfterCollapse >= allItems.length) {
        process.env.NODE_ENV !== "production" ? warning(false, ['Material-UI: you have provided an invalid combination of properties to the Breadcrumbs.', // eslint-disable-next-line max-len
        "itemsAfterCollapse={".concat(itemsAfterCollapse, "} + itemsBeforeCollapse={").concat(itemsBeforeCollapse, "} >= maxItems={").concat(maxItems, "}")].join('\n')) : void 0;
        return allItems;
      }

      return [].concat(_toConsumableArray(allItems.slice(0, itemsBeforeCollapse)), [React.createElement(BreadcrumbCollapsed, {
        key: "ellipsis",
        onClick: this.handleClickExpand
      })], _toConsumableArray(allItems.slice(allItems.length - itemsAfterCollapse, allItems.length)));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          classes = _this$props2.classes,
          classNameProp = _this$props2.className,
          Component = _this$props2.component,
          itemsAfterCollapse = _this$props2.itemsAfterCollapse,
          itemsBeforeCollapse = _this$props2.itemsBeforeCollapse,
          maxItems = _this$props2.maxItems,
          separator = _this$props2.separator,
          other = _objectWithoutProperties(_this$props2, ["children", "classes", "className", "component", "itemsAfterCollapse", "itemsBeforeCollapse", "maxItems", "separator"]);

      var allItems = React.Children.toArray(children).filter(function (child) {
        return React.isValidElement(child);
      }).map(function (child, index) {
        return React.createElement("li", {
          className: classes.li,
          key: String(index)
        }, child);
      });
      return React.createElement(Typography, _extends({
        component: Component,
        color: "textSecondary",
        className: clsx(classes.root, classNameProp)
      }, other), React.createElement("ol", {
        className: classes.ol
      }, this.insertSeparators(this.state.expanded || maxItems && allItems.length <= maxItems ? allItems : this.renderItemsBeforeAndAfter(allItems))));
    }
  }]);

  return Breadcrumbs;
}(React.Component);

process.env.NODE_ENV !== "production" ? Breadcrumbs.propTypes = {
  /**
   * The breadcrumb children.
   */
  children: PropTypes.node.isRequired,

  /**
   * Override or extend the styles applied to the component.
   * See [CSS API](#css) below for more details.
   */
  classes: PropTypes.object.isRequired,

  /**
   * @ignore
   */
  className: PropTypes.string,

  /**
   * The component used for the root node.
   * Either a string to use a DOM element or a component.
   * By default, it maps the variant to a good default headline component.
   */
  component: componentPropType,

  /**
   * If max items is exceeded, the number of items to show after the ellipsis.
   */
  itemsAfterCollapse: PropTypes.number,

  /**
   * If max items is exceeded, the number of items to show before the ellipsis.
   */
  itemsBeforeCollapse: PropTypes.number,

  /**
   * Specifies the maximum number of breadcrumbs to display. When there are more
   * than the maximum number, only the first and last will be shown, with an
   * ellipsis in between.
   */
  maxItems: PropTypes.number,

  /**
   * Custom separator node.
   */
  separator: PropTypes.node
} : void 0;
Breadcrumbs.defaultProps = {
  component: 'nav',
  itemsAfterCollapse: 1,
  itemsBeforeCollapse: 1,
  maxItems: 8,
  separator: '/'
};
export default withStyles(styles, {
  name: 'MuiBreadcrumbs'
})(Breadcrumbs);
